README.md
# Helix peformance testing

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-perf.svg)](https://codecov.io/gh/adobe/helix-perf)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-perf.svg)](https://circleci.com/gh/adobe/helix-perf)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-perf.svg)](https://github.com/adobe/helix-perf/blob/master/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-perf.svg)](https://github.com/adobe/helix-perf/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-perf.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-perf) 


> Microservice for getting Project Helix performance results

This microservice runs performance tests on a customer website using Calibreapp, but only if the customer has a valid Project Helix account.

## Usage

Send a POST request with following (`Content-Type: application/json`-encoded) body parameters to `https://adobeioruntime.net/api/v1/web/helix/helix-services/perf@v1`:

* `service`: the service ID of your Fastly service config
* `token`: a Fastly authentication token that has `global` permission on the service config
* `tests`: an array of test specs, i.e. objects with following properties
  * `url`: the URL to test
  * `location`: the remote location to test from
  * `device`: the device emulate during the test
  * `connection`: the connection throttling settings
  * `strain`: the strain or variant to test

## Developing Helix Perf

You need `node>=8.0.0` and `npm>=5.4.0`. Follow the typical `npm install`, `npm test` workflow.

Contributions are highly welcome.

## Deploying Helix Perf

Deploying Helix Perf requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

Run `npm run deploy` to do a one-shot deploment of Helix Perf. All commits to master that pass the testing will be deployed automatically.


CHANGELOG.md
## [1.1.151](https://github.com/adobe/helix-perf/compare/v1.1.150...v1.1.151) (2021-04-17)


### Bug Fixes

* **deps:** update dependency calibre to v3.2.2 ([8cec1be](https://github.com/adobe/helix-perf/commit/8cec1be1c1936ab6e87df071c68c4e0acece1d2c))

## [1.1.150](https://github.com/adobe/helix-perf/compare/v1.1.149...v1.1.150) (2021-04-12)


### Bug Fixes

* **deps:** update dependency @adobe/fastly-native-promises to v1.23.0 ([42b1bb2](https://github.com/adobe/helix-perf/commit/42b1bb2600b39948770159b20b703479672cf8c5))

## [1.1.149](https://github.com/adobe/helix-perf/compare/v1.1.148...v1.1.149) (2021-04-12)


### Bug Fixes

* **deps:** update dependency @adobe/openwhisk-action-logger to v2.4.4 ([#460](https://github.com/adobe/helix-perf/issues/460)) ([55902c9](https://github.com/adobe/helix-perf/commit/55902c9c348b251a79ea9092602a99acd0ff122c))

## [1.1.148](https://github.com/adobe/helix-perf/compare/v1.1.147...v1.1.148) (2021-04-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.6 ([#459](https://github.com/adobe/helix-perf/issues/459)) ([0933f24](https://github.com/adobe/helix-perf/commit/0933f24896ce8f2d146fc7e4f35d49a321a2d9de))

## [1.1.147](https://github.com/adobe/helix-perf/compare/v1.1.146...v1.1.147) (2021-04-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.6.4 ([#457](https://github.com/adobe/helix-perf/issues/457)) ([45f6847](https://github.com/adobe/helix-perf/commit/45f68471c3e514dbed0647de3600deeb539d2767))

## [1.1.146](https://github.com/adobe/helix-perf/compare/v1.1.145...v1.1.146) (2021-04-09)


### Bug Fixes

* **deps:** update adobe fixes ([43ee0b7](https://github.com/adobe/helix-perf/commit/43ee0b741a866d89958a7549acc4dda0861724aa))

## [1.1.145](https://github.com/adobe/helix-perf/compare/v1.1.144...v1.1.145) (2021-04-08)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.6.2 ([3d725e0](https://github.com/adobe/helix-perf/commit/3d725e0c81c89ffd2b77d0c4661bca23c35453f9))

## [1.1.144](https://github.com/adobe/helix-perf/compare/v1.1.143...v1.1.144) (2021-04-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.5 ([11e8648](https://github.com/adobe/helix-perf/commit/11e8648028fc3b1dfd82445b68c353098623c4a6))

## [1.1.143](https://github.com/adobe/helix-perf/compare/v1.1.142...v1.1.143) (2021-04-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.4 ([1de3d04](https://github.com/adobe/helix-perf/commit/1de3d04ccf95b2d977f5f469699160067e92a323))

## [1.1.142](https://github.com/adobe/helix-perf/compare/v1.1.141...v1.1.142) (2021-03-30)


### Bug Fixes

* **deps:** update adobe fixes ([6cc2c94](https://github.com/adobe/helix-perf/commit/6cc2c9435507a77a9a2e8c83cdde2567d7ad16ce))

## [1.1.141](https://github.com/adobe/helix-perf/compare/v1.1.140...v1.1.141) (2021-03-26)


### Bug Fixes

* **deps:** update adobe fixes ([26f2b0c](https://github.com/adobe/helix-perf/commit/26f2b0cd55ab07f1b64ef6a1c699a90c72cb5eea))

## [1.1.140](https://github.com/adobe/helix-perf/compare/v1.1.139...v1.1.140) (2021-03-26)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.2 ([#441](https://github.com/adobe/helix-perf/issues/441)) ([0f22576](https://github.com/adobe/helix-perf/commit/0f2257608f42d6157791f87f6e01d2571ca5890b))

## [1.1.139](https://github.com/adobe/helix-perf/compare/v1.1.138...v1.1.139) (2021-03-26)


### Bug Fixes

* **deps:** update adobe fixes ([1ebb1dd](https://github.com/adobe/helix-perf/commit/1ebb1ddc9da9be98aa6bd81b632e9c6af4add53e))

## [1.1.138](https://github.com/adobe/helix-perf/compare/v1.1.137...v1.1.138) (2021-03-23)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.1 ([#439](https://github.com/adobe/helix-perf/issues/439)) ([90c4cb5](https://github.com/adobe/helix-perf/commit/90c4cb5e763679bea30c623e2e3bdf2c395cd648))

## [1.1.137](https://github.com/adobe/helix-perf/compare/v1.1.136...v1.1.137) (2021-03-10)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.0 ([#431](https://github.com/adobe/helix-perf/issues/431)) ([08eb6d6](https://github.com/adobe/helix-perf/commit/08eb6d6f241fb977c7acf0c1258eefcd93d8c54d))

## [1.1.136](https://github.com/adobe/helix-perf/compare/v1.1.135...v1.1.136) (2021-03-09)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.0.9 ([#430](https://github.com/adobe/helix-perf/issues/430)) ([6ef7f37](https://github.com/adobe/helix-perf/commit/6ef7f3728a420fe4c72acbd1566c6afee74c054b))

## [1.1.135](https://github.com/adobe/helix-perf/compare/v1.1.134...v1.1.135) (2021-03-09)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.0.8 ([9875a65](https://github.com/adobe/helix-perf/commit/9875a65eaff35f64cff815f6d8bb9cd8547e9ae2))

## [1.1.134](https://github.com/adobe/helix-perf/compare/v1.1.133...v1.1.134) (2021-03-08)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.0.7 ([18ce549](https://github.com/adobe/helix-perf/commit/18ce549a4f6472d1448080e6bff6cbc7bc0e3029))

## [1.1.133](https://github.com/adobe/helix-perf/compare/v1.1.132...v1.1.133) (2021-03-08)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.6.0 ([#427](https://github.com/adobe/helix-perf/issues/427)) ([4095826](https://github.com/adobe/helix-perf/commit/4095826370f92f630f54a9bbb2df5ddecc240299))

## [1.1.132](https://github.com/adobe/helix-perf/compare/v1.1.131...v1.1.132) (2021-03-05)


### Bug Fixes

* **deps:** update adobe fixes ([#424](https://github.com/adobe/helix-perf/issues/424)) ([65bab01](https://github.com/adobe/helix-perf/commit/65bab01b146b4b0ff5590b2dee846f0a2eea53b9))

## [1.1.131](https://github.com/adobe/helix-perf/compare/v1.1.130...v1.1.131) (2021-03-02)


### Bug Fixes

* **deps:** update dependency @adobe/fastly-native-promises to v1.22.1 ([2291fc2](https://github.com/adobe/helix-perf/commit/2291fc2f5b32d603e4d8d663055094b793d77ce5))


test/condit.js
/*
 * Copyright 2018 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
/* eslint-env mocha */

function condit(name, condition, mochafn) {
  if (condition()) {
    return it(name, mochafn);
  }
  return it.skip(`${name} (${condition.description || 'condition not met'})`, mochafn);
}

condit.hasenv = (name) => {
  class fn {
    constructor() {
      return !!process.env[name];
    }
  }
  fn.description = `env var ${name} must be set`;
  return fn;
};

condit.hasenvs = (names) => {
  const fn = function envs() {
    return names.reduce((p, c) => p && !!process.env[c], true);
  };

  fn.description = `env vars ${names.join(', ')} must be set`;
  return fn;
};

module.exports = condit;


test/index.test.js
/*
 * Copyright 2018 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';

const assert = require('assert');
const proxyquire = require('proxyquire').noCallThru();
const sinon = require('sinon');

describe('Mocked Tests', () => {
  const example = {
    uuid: '170b278',
    url: 'https://debug.primordialsoup.life/develop/',
    formattedTestUrl: 'https://calibreapp.com/tests/170b278/4161820',
    status: 'completed',
    updatedAt: '2018-08-29T13:00:08Z',
    metrics: [
      { name: 'speed_index', label: 'Speed Index', value: 1233 },
      {
        name: 'visually_complete',
        label: 'Visually Complete',
        value: 1233,
      },
      {
        name: 'visually_complete_85',
        label: '85% Visually Complete',
        value: 1233,
      },
      {
        name: 'lighthouse-seo-score',
        label: 'Lighthouse SEO Score',
        value: 44,
      },
      {
        name: 'lighthouse-best-practices-score',
        label: 'Lighthouse Best Practices Score',
        value: 87,
      },
      {
        name: 'lighthouse-accessibility-score',
        label: 'Lighthouse Accessibility Score',
        value: 50,
      },
      {
        name: 'lighthouse-performance-score',
        label: 'Lighthouse Performance Score',
        value: 99,
      },
      {
        name: 'lighthouse-pwa-score',
        label: 'Lighthouse Progressive Web App Score',
        value: 36,
      },
      {
        name: 'js-parse-compile',
        label: 'JS Parse & Compile',
        value: 0,
      },
      {
        name: 'time-to-first-byte',
        label: 'Time to First Byte',
        value: 1064,
      },
      {
        name: 'first-contentful-paint',
        label: 'First Contentful Paint',
        value: 1218,
      },
      {
        name: 'first-meaningful-paint',
        label: 'First Meaningful Paint',
        value: 1218,
      },
      { name: 'firstRender', label: 'First Paint', value: 1218 },
      { name: 'dom-size', label: 'DOM Element Count', value: 3 },
      {
        name: 'estimated-input-latency',
        label: 'Estimated input latency',
        value: 16,
      },
      {
        name: 'consistently-interactive',
        label: 'Time to Interactive',
        value: 1218,
      },
      {
        name: 'first-interactive',
        label: 'First CPU Idle',
        value: 1218,
      },
      {
        name: 'html_body_size_in_bytes',
        label: 'Total HTML size in bytes',
        value: 80,
      },
      {
        name: 'html_size_in_bytes',
        label: 'Total HTML transferred',
        value: 239,
      },
      { name: 'page_wait_timing', label: 'Response time', value: 1236 },
      {
        name: 'page_size_in_bytes',
        label: 'Total Page transferred',
        value: 239,
      },
      {
        name: 'page_body_size_in_bytes',
        label: 'Total Page size in bytes',
        value: 80,
      },
      { name: 'asset_count', label: 'Number of requests', value: 1 },
      { name: 'onload', label: 'onLoad', value: 1196 },
      { name: 'oncontentload', label: 'onContentLoad', value: 1198 },
    ],
    device: { title: 'Motorola Moto G4' },
    connection: { title: 'Regular 3G' },
    location: { name: 'London, United Kingdom', emoji: '🇬🇧' },
  };

  const fakecreate = sinon.fake.resolves({ uuid: '170b278' });
  const fakewait = sinon.fake.resolves(example);
  const prf = proxyquire('../src/perf', {
    calibre: {
      Test: {
        create: fakecreate,
        waitForTest: fakewait,
      },
    },
    '@adobe/fastly-native-promises': () => ({
      readVersions: sinon.fake.resolves(true),
    }),
  });

  const { main: perf } = proxyquire('../src/index', {
    './perf': prf,
  });

  it('Works without arguments', async () => {
    const result = await perf({ CALIBRE_AUTH: 'FAKE' });
    assert.deepEqual(result.statusCode, 200);
  });

  it('Can run tests', async () => {
    const result = await perf({
      __ow_method: 'post',
      CALIBRE_AUTH: 'FAKE',
      tests: [
        {
          url: 'https://www.project-helix.io',
          location: 'London',
          device: 'MotorolaMotoG4',
          connection: 'regular3G',
          strain: 'default',
        },
        {
          url: 'https://www.project-helix.io',
          location: 'London',
          device: 'MotorolaMotoG4',
          connection: 'regular3G',
          strain: 'default',
        },
      ],
    });
    assert.deepEqual(result.body, ['170b278', '170b278']);
  });

  it('Can retrieve tests', async () => {
    const result = await perf({
      __ow_method: 'post',
      CALIBRE_AUTH: 'FAKE',
      tests: ['170b278', '170b278'],
    });
    assert.deepEqual(result.body, [{ uuid: '170b278', ...example },
      {
        uuid: '170b278', ...example,
      }]);
  });
});


test/integration.test.js
/*
 * Copyright 2018 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';

const assert = require('assert');
const condit = require('./condit');
const perf = require('../src/perf');

describe('Integration Tests', () => {
  it('Rejects attempts to consume service without authentication', (done) => {
    perf({ __ow_method: 'post' }).then((result) => {
      assert.equal(result.statusCode, 401);
      done();
    }).catch(() => done);
  });

  condit(
    'Fail when called with invalid configuration',
    condit.hasenvs(['HLX_CALIBRE_AUTH', 'HLX_FASTLY_AUTH', 'HLX_FASTLY_NAMESPACE']),
    async () => {
      const results = await perf({
        __ow_method: 'post',
        CALIBRE_API_TOKEN: process.env.HLX_CALIBRE_AUTH,
        service: process.env.HLX_FASTLY_NAMESPACE,
        token: process.env.HLX_FASTLY_AUTH,
        tests: [
          {
            url: 'https://www.project-helix.io',
            location: 'London',
            device: 'Nokia7110',
            connection: 'regular3G',
            strain: 'default',
          },
          {
            url: 'https://www.adobe.io',
            location: 'London',
            device: 'Nokia7110',
            connection: 'regular3G',
            strain: 'default',
          },
        ],
      });

      assert.equal(results.statusCode, 500);
    },
  ).timeout(1000 * 60 * 9);

  condit(
    'Fail to retrieve when called with invalid configuration',
    condit.hasenvs(['HLX_CALIBRE_AUTH', 'HLX_FASTLY_AUTH', 'HLX_FASTLY_NAMESPACE']),
    async () => {
      const results = await perf({
        __ow_method: 'post',
        CALIBRE_API_TOKEN: process.env.HLX_CALIBRE_AUTH,
        service: process.env.HLX_FASTLY_NAMESPACE,
        token: process.env.HLX_FASTLY_AUTH,
        tests: [
          'nope', 'doesntexit',
        ],
      });

      assert.equal(results.statusCode, 500);
    },
  ).timeout(1000 * 60 * 9);

  condit(
    'Fail when called with invalid credentials',
    condit.hasenvs(['HLX_CALIBRE_AUTH', 'HLX_FASTLY_AUTH', 'HLX_FASTLY_NAMESPACE']),
    async () => {
      const results = await perf({
        __ow_method: 'post',
        CALIBRE_API_TOKEN: process.env.HLX_CALIBRE_AUTH,
        service: 'wrong',
        token: 'fake',
        tests: [
          {
            url: 'https://www.project-helix.io',
            location: 'London',
            device: 'Nokia7110',
            connection: 'regular3G',
            strain: 'default',
          },
          {
            url: 'https://www.adobe.io',
            location: 'London',
            device: 'Nokia7110',
            connection: 'regular3G',
            strain: 'default',
          },
        ],
      });

      assert.equal(results.statusCode, 401);
    },
  ).timeout(1000 * 60 * 9);

  condit(
    'Fail when retrieving with invalid credentials',
    condit.hasenvs(['HLX_CALIBRE_AUTH', 'HLX_FASTLY_AUTH', 'HLX_FASTLY_NAMESPACE']),
    async () => {
      const results = await perf({
        __ow_method: 'post',
        CALIBRE_API_TOKEN: process.env.HLX_CALIBRE_AUTH,
        service: 'wrong',
        token: 'fake',
        tests: ['one', 'two'],
      });

      assert.equal(results.statusCode, 401);
    },
  ).timeout(1000 * 60 * 9);

  condit(
    'Retrieve Performance results from Calibre',
    condit.hasenvs(['HLX_CALIBRE_AUTH', 'HLX_FASTLY_AUTH', 'HLX_FASTLY_NAMESPACE']),
    async () => {
      const schedule = await perf({
        __ow_method: 'post',
        CALIBRE_API_TOKEN: process.env.HLX_CALIBRE_AUTH,
        service: process.env.HLX_FASTLY_NAMESPACE,
        token: process.env.HLX_FASTLY_AUTH,
        tests: [
          {
            url: 'https://www.project-helix.io',
            location: 'London',
            device: 'MotorolaMotoG4',
            connection: 'regular3G',
            strain: 'default',
          },
          {
            url: 'https://www.adobe.io',
            location: 'London',
            device: 'MotorolaMotoG4',
            connection: 'regular3G',
            strain: 'default',
          },
        ],
      });

      assert.equal(schedule.body.length, 2);

      let i = 0;
      // eslint-disable-next-line no-plusplus
      while (i++ < 100) {
        // eslint-disable-next-line no-await-in-loop
        const results = await perf({
          __ow_method: 'post',
          tests: schedule.body,
          CALIBRE_API_TOKEN: process.env.HLX_CALIBRE_AUTH,
          service: process.env.HLX_FASTLY_NAMESPACE,
          token: process.env.HLX_FASTLY_AUTH,
        });

        assert.equal(results.body.length, 2);
        if (results.body.reduce((p, result) => p && typeof result === 'object', true)) {
          const r1 = results.body[0];
          assert.equal(r1.url, 'https://www.project-helix.io');
          assert.equal(typeof r1.uuid, 'string');
          assert.ok(Array.isArray(r1.metrics));
          assert.equal(typeof r1.metrics[0].value, 'number');

          break;
        } else {
          // eslint-disable-next-line no-console
          console.log(`results have not yet arrived after ${i} iterations`);
        }
      }
    },
  ).timeout(1000 * 60 * 9);
});


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const { wrap: status } = require('@adobe/helix-status');
const { wrap } = require('@adobe/openwhisk-action-utils');
const { logger } = require('@adobe/openwhisk-action-logger');
const { epsagon } = require('@adobe/helix-epsagon');
const perf = require('./perf.js');

/**
 * Main function called by the openwhisk invoker.
 * @param params Action params
 * @returns {Promise<*>} The response
 */
module.exports.main = wrap(perf)
  .with(epsagon)
  // TODO: enable checks again, once we have support for more elaborate requests.
  // TODO: see: https://github.com/adobe/helix-status/issues/48
  // .with(status, { calibre: 'https://calibreapp.com/graphql' })
  .with(status)
  .with(logger.trace)
  .with(logger);


src/perf.js
/*
 * Copyright 2018 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const URI = require('uri-js');
const fastly = require('@adobe/fastly-native-promises');

function setENVtoken(token) {
  process.env.CALIBRE_API_TOKEN = token;
}

function init(token) {
  setENVtoken(token);
  let calibre;
  return () => {
    if (calibre) {
      return calibre;
    }
    // calibre loads the environment variables at require-time
    // so we delay loading until called.
    /* eslint-disable global-require */
    calibre = require('calibre');
    return calibre;
  };
}

function test(calibre, {
  url, location, device, connection, strain,
}) {
  return calibre.Test.create({
    url,
    location,
    device,
    connection,
    cookies: [
      {
        name: 'X-Strain',
        value: strain,
        secure: true,
        httpOnly: true,
        domain: URI.parse(url).host,
      },
    ],
  }).then(async ({ uuid }) => uuid);
}

async function result(calibre, uuid) {
  const nothing = new Promise((resolve) => {
    // maximum response time for OpenWhisk is 60 seconds,
    // so let's wait no longer than 45 seconds.
    setTimeout(resolve, 1000 * 45, uuid);
  });

  const testresult = calibre.Test.waitForTest(uuid);

  const res = await Promise.race([nothing, testresult]);
  if (typeof res === 'string') {
    return uuid;
  }
  return res;
}

async function perf({
  CALIBRE_API_TOKEN, service, token, tests = [],
}) {
  const getcalibre = init(CALIBRE_API_TOKEN);
  if (
    tests.length > 0
    && tests.reduce((p, uuid) => p && typeof uuid === 'string', true)
  ) {
    return fastly(token, service)
      .readVersions()
      .then(() => Promise.all(tests.map((uuid) => result(getcalibre(), uuid)))
        .then((res) => ({
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json',
          },
          body: res,
        }))
        .catch((e) => ({
          statusCode: 500,
          body: `Unable to retrieve test results ${e}`,
        })))
      .catch(() => ({
        statusCode: 401,
        body: 'Invalid credentials.',
      }));
    // eslint-disable-next-line camelcase
  }
  return fastly(token, service)
    .readVersions()
    .then(() => Promise.all(tests.map((spec) => test(getcalibre(), spec)))
      .then((res) => ({
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json',
        },
        body: res,
      }))
      .catch((e) => ({
        statusCode: 500,
        body: `Unable to perfom test ${e}`,
      })))
    .catch(() => ({
      statusCode: 401,
      body: 'Invalid credentials.',
    }));
}

module.exports = perf;
